File: cpp/022.GenerateParentheses/GenerateParentheses.cpp
/*
解题思路：
    回溯
*/

// Runtime Beats 100.00%

#include <cassert>
#include <iostream>
#include <string>
#include <vector>

class Solution {
public:
    std::vector<std::string> generateParenthesis(int n) {
        std::string path;
        backtrack(0, 0, n, path);
        return results;
    }

    void backtrack(int left, int right, int n, std::string& path) {
        if (left == n && right == n) {
            results.emplace_back(path);
            return;
        }

        if (left < n) {
            path.push_back('(');
            backtrack(left + 1, right, n, path);
            path.pop_back();
        }

        if (right < left) {
            path.push_back(')');
            backtrack(left, right + 1, n, path);
            path.pop_back();
        }
    }

private:
    std::vector<std::string> results;
};

void test1() {
    std::vector<std::string> results{"((()))", "(()())", "(())()", "()(())", "()()()"};
    auto n = 3;
    Solution s;
    assert(s.generateParenthesis(n) == results);
}

void test2() {
    std::vector<std::string> results{"()"};
    auto n = 1;
    Solution s;
    assert(s.generateParenthesis(n) == results);
}

int main() {
    test1();
    test2();
    return 0;
}

File: python3/022.GenerateParentheses/GenerateParentheses.py
# 解题思路：
#     回溯

# Runtime Beats 31.69%

from typing import List


class Solution:

    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.backtrack(0, 0, n, '', result)
        return result

    def backtrack(self, left, right, n, s, result):
        if left == n and right == n:
            result.append(s)
            return

        if left < n:
            self.backtrack(left + 1, right, n, s + '(', result)

        if right < left:
            self.backtrack(left, right + 1, n, s + ')', result)


def test1():
    result = ['((()))', '(()())', '(())()', '()(())', '()()()']
    s = Solution()
    assert s.generateParenthesis(3) == result


def test2():
    result = ['()']
    s = Solution()
    assert s.generateParenthesis(1) == result


test1()
test2()
